alg/viewshed/cumulative_viewshed.cpp

```diff
@@ -0,0 +1,196 @@
+    m_observerQueue.done();
+
+    // Run executors.
+    const int numThreads = NUM_JOBS;
+    std::atomic<bool> err = false;
+    std::atomic<int> running = numThreads;
+    CPLWorkerThreadPool executorPool(numThreads);
+    for (int i = 0; i < numThreads; ++i)
+        executorPool.SubmitJob([this, &srcFilename, &err, &running]
+                               { runExecutor(srcFilename, err, running); });
+
+    // Run combiners that create 8-bit sums of executor jobs.
+    CPLWorkerThreadPool combinerPool(numThreads);
+    std::vector<Combiner> combiners(numThreads,
+                                    Combiner(m_datasetQueue, m_bufQueue));
+    for (Combiner &c : combiners)
+        combinerPool.SubmitJob([&c] { c.run(); });
+
+    // Run 32-bit rollup job.
+    std::thread sum([this] { rollupRasters(); });
+
+    // When the combiner jobs are done, all the data is in the buf queue.
+    combinerPool.WaitCompletion();
+    if (m_datasetQueue.isStopped())
+        return false;
+    m_bufQueue.done();
+
+    // Wait for finalBuf to be fully filled. Then scale the output data.
+    sum.join();
+    scaleOutput();
+    bool ok = writeOutput(*pSrcBand);
+
+    executorPool.WaitCompletion();
+    (void)pfnProgress;
+    (void)pProgressArg;
+    return ok;
+}
+
+void Cumulative::runExecutor(const std::string &srcFilename,
+                             std::atomic<bool> &err, std::atomic<int> &running)
+{
+    Location loc;
+    while (!err && m_observerQueue.pop(loc))
+    {
+        DatasetPtr srcDs(GDALDataset::Open(srcFilename.c_str(), GA_ReadOnly));
+        if (!srcDs)
+        {
+            err = true;
+            break;
+        }
+
+        GDALDriver *memDriver = GetGDALDriverManager()->GetDriverByName("MEM");
+        DatasetPtr dstDs(memDriver->Create(
+            "", m_extent.xSize(), m_extent.ySize(), 1, GDT_Byte, nullptr));
+        ViewshedExecutor executor(*srcDs->GetRasterBand(1),
+                                  *dstDs->GetRasterBand(1), loc.x, loc.y,
+                                  m_extent, m_extent, m_opts);
+
+        if (!executor.run())
+            err = true;  // Signal other threads to stop.
+        else
+            m_datasetQueue.push(std::move(dstDs));
+    }
+
+    // Job done. Set the output queue state.  If all the executor jobs have completed,
+    // set the dataset output queue done.
+    if (err)
+        m_datasetQueue.stop();
+    else
+    {
+        running--;
+        if (!running)
+            m_datasetQueue.done();
+    }
+}
+
+// Add 8-bit rasters into the 32-bit raster buffer.
+void Cumulative::rollupRasters()
+{
+    Buf8 buf;
+
+    m_finalBuf.resize(m_extent.size());
+    while (m_bufQueue.pop(buf))
+        for (size_t i = 0; i < m_finalBuf.size(); ++i)
+            m_finalBuf[i] += buf[i];
+}
+
+void Cumulative::scaleOutput()
+{
+    uint32_t m = 0;  // This gathers all the bits set.
+    for (uint32_t &val : m_finalBuf)
+        m = std::max(val, m);
+
+    double factor =
```

Comment:
is there an chance that m could be 0 and a division by zero occuring ?

Reply:
Fixed.